From e36d435bdd42214be2d7824c224a9fb830e2f6d3 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 06 Jul 2021 20:31:28 +0300
Subject: [PATCH] [SW-22550] ppboot: greco: pcie drain enable

Currently while a fw reset is triggered, the vuart daemon is unable
to retrieve the data and status which is a part of PCIE_ELBI
block.

This patch add changes to enable pcie drain control and set timeout
of 1000 cycles which invalidates incomplete transaction after timeout.
Upon performing above configuration the complete logs are visible
on the vuart daemon console.

Change-Id: Iee042d3b7ff2faa717dab36729a49eef7ef469d8
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/arch/greco/common.c b/arch/greco/common.c
index d813bfe..0f74f3d 100644
--- a/arch/greco/common.c
+++ b/arch/greco/common.c
@@ -16,6 +16,7 @@
 #include <pcie_aux_regs.h>
 #include <pcie_core_regs.h>
 #include <pcie_wrap_regs.h>
+#include <pcie_wrap_masks.h>
 #include <pcie_phy_regs.h>
 #include <pcie_phy_masks.h>
 #include <pcie_dbi_regs.h>
@@ -563,9 +564,27 @@
 	regmap_writel(val, mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
 }
 
+static void axi_enable_pcie_drain(void)
+{
+	u32 val;
+
+	/* Set timeout and enable drain */
+	regmap_writel(PCIE_AXI_DRAIN_TIMEOUT, mmPCIE_WRAP_HBW_DRAIN_TIMEOUT);
+	regmap_writel(PCIE_AXI_DRAIN_TIMEOUT, mmPCIE_WRAP_LBW_DRAIN_TIMEOUT);
+
+	val = regmap_readl(mmPCIE_WRAP_HBW_DRAIN_CFG);
+	val |= BIT(PCIE_WRAP_HBW_DRAIN_CFG_EN_SHIFT);
+	regmap_writel(val, mmPCIE_WRAP_HBW_DRAIN_CFG);
+
+	val = regmap_readl(mmPCIE_WRAP_LBW_DRAIN_CFG);
+	val |= BIT(PCIE_WRAP_LBW_DRAIN_CFG_EN_SHIFT);
+	regmap_writel(val, mmPCIE_WRAP_LBW_DRAIN_CFG);
+}
+
 void hl_plat_early_init(void)
 {
 	axi_enable_drain();
+	axi_enable_pcie_drain();
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
 }
 
diff --git a/arch/greco/common.h b/arch/greco/common.h
index 819ed3e..8f3131f 100644
--- a/arch/greco/common.h
+++ b/arch/greco/common.h
@@ -83,4 +83,5 @@
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_SEL_G4_CONF	0xFF
 
 #define PSOC_AXI_DRAIN_TIMEOUT				0x3E8 /* 1000 cycles */
+#define PCIE_AXI_DRAIN_TIMEOUT				PSOC_AXI_DRAIN_TIMEOUT
 #endif /* COMMON_GRECO_H_ */
